scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,359 @@
+    """
+    Add an export.
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to add
+    :type export_name: string (unicode)
+    """
+
+    if not export_name or '/' in export_name:
+        raise ExportException('Invalid export name')
+
+    export_point = os.path.join(root_export, export_name)
+
+    # Copied from Manila generic driver
+    create_cmd = [
+        'net', 'conf', 'addshare', export_name, export_point,
+        'writeable=y', 'guest_ok=y',
+    ]
+    parameters = {
+        'browseable': 'yes',
+        '\"create mask\"': '0755',
+        '\"hosts deny\"': '0.0.0.0/0',  # deny all by default
+        '\"hosts allow\"': '127.0.0.1',
+        '\"read only\"': 'no',
+    }
+    set_of_commands = [':', ]  # : is just placeholder
+    for param, value in parameters.items():
+        # These are combined in one list to run in one process
+        # instead of big chain of one action calls.
+        set_of_commands.extend(['&&', 'net', 'conf', 'setparm',
+                                export_name, param, value])
+
+    with utils.elevated_privileges():
+        try:
+            os.mkdir(export_point)
+        except OSError as exc:
+            if exc.errno != errno.EEXIST:
+                raise
+            else:
+                log.debug("The share/directory %s already exists on SOFS",
+                          export_name)
+                exports = _get_defined_exports()
+                if export_name in exports:
+                    msg = ("Share '{0:s}' already defined in Samba "
+                           "registry.".format(export_name))
+                    raise ExportAlreadyExists(msg)
+
+        os.chmod(export_point, 0o0777)
+
+        subprocess.check_call(create_cmd)
+        subprocess.check_call(' '.join(set_of_commands), shell=True)
+
+
+@ensure_environment
+@ensure_export_exists
+def wipe_export(root_export, export_name, exports):
+    """
+    Remove an export.
+
+    The export point is not actually removed, but renamed with the prefix
+    "TRASH-".
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to remove
+    :type export_name: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    """
+
+    export = exports[export_name]
+    export_path = os.path.join(root_export, export_name)
+    if export['hosts allow'] and export['hosts allow'] != '127.0.0.1':
+        raise ExportHasGrantsException('Unable to remove export with grants')
+
+    # We need to introduce a "variable" part (i.e the UNIX timestamp)
+    # in case an export with the same name is deleted twice
+    tombstone = 'TRASH-{0:s}-{1:d}'.format(export_name, int(time.time()))
```

Comment:
We could use a uuid if all we care about is uniqueness. However I think a normal timestamp can be an interesting piece of information for the administrator, and if so include a timezone as we are in a distributed system.